keps/159-scheduler-simulator-operator/README.md

```diff
@@ -0,0 +1,118 @@
+# KEP-159: scheduler simulator operator
+
+## Summary
+
+A new custom resource to represent a simulator itself and a custom controller to manage it are introduced.
+
+## Motivation
+
+With this, we can easily create/use the simulator as sandbox for the simulator by creating the Simulator custom resource. 
+
+This feature also makes it easier to handle simulators from other CRD or controllers. 
+They can gets the information for the simulator easily by accessint the `Simulator` resource.
+
+### Goals
+
+- Users can define simulators with Simulator CRD.
+- The controller will manage(create, edit, delete) simulators.
+
+### Non-Goals
+
+Create new `Simulator` from Web UI. (may be implemented in the future, but out of scope of this proposal.)
+
+## Proposal
+
+### Implementation design details
+
+#### Required prior knowledge
+
+The current simulator is built in a single binary: the binary contains kube-apiserver, kube-scheduler and the simulator backend (which provides the web UI specific feature.)
+
+And, the simulator can be configured via environment variables
+
+#### Simulator CRD
+
+The CRD `Simulator` will be applied to kube-apiserver in user's cluster. (not to simulator's kube-apiserver.) 
+
+It will create the Pod for the simulator in the same namespace.
+
+```go
+type Simulator struct {
+  metav1.TypeMeta 
+  metav1.ObjectMeta // namespaced.
+
+  Spec SimulatorSpec
+  Status SimulatorStatus
+}
+
+// SimulatorSpec has the spec for the simulator.
+// All fields can be updated (but a Pod may be re-created.)
```

Comment:
Don't we need to handle existing(finished/running/pending) `Scenario` resources inside of the sumulator??  I think we should wait for users to clean up `Scenario` in the simulator. What do you think?

Reply:
Good question.

> we should wait for users to clean up Scenario in the simulator.

What did you mean by "clean up"? Let users delete all Scenario resources in the simulator? So, I don't want to let users do something for restarting the simulator Pod (since it's hard for users to realize they need to do that to kick restarting), but want the `Simulator` controller to do all the things for restarting the simulator Pod. 

How about this:
- implementing a way (endpoint on the simulator server?) that tells the Scenario controller that the Simulator controller wants to shut down the Pod.
  - when the Scenario controller moves FakeTime forward, it checks if the shutdown notification has come or not. If it has, the Scenario controller changes the Scenario's phase to "Paused" instead of continuing running the Scenario.
- setting `preStop` to notify shutdown to the Scenario controller by the above way.

Reply:
I think we can put this in "Future work", and until implementing them, the controller will behave like this: when the controller is started and finds the Scenario which phase is "Running", the controller just changes the status "Failed" with the `.status.message` like "the controller restarted while the Scenario was running".

Reply:
> Let users delete all Scenario resources in the simulator?

Yeah, I thought so.

> How about this:

that sounds reasonable about handling the running Scenario. How about completed scenarios? If a Simulator CR (i.e. simulator pod) hosts multiple (say a lot of) completed Scenario CRs and what if a user updated or deleted the Simulator CR accidentally??  I think we should let users resurrect Scenario results in some way.

Reply:
>  How about completed scenarios?

We don't need to do anything. The etcd has the data for completed scenario.

> what if a user updated or deleted the Simulator CR accidentally??

Could you elaborate on that? The system should not resurrect what our users have updated/deleted.

Reply:
> I think we can put this in "Future work"

I'll write this in https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/186

Reply:
Added the note for this on https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/186